### PR TITLE
Fix S1172 FP: Parameter captured in a referenced generic local function

### LINKED ISSUE #4704: Fix S1172 FP: Parameter captured in a referenced generic local function
### Description

The S1172 warning is shown for a parameter captured by a generic local function.

### Repro steps

```csharp
        private static void ConfigureAndValidateSettings(this IServiceCollection services, IConfiguration configuration)
        {
            ConfigureAndValidate<MySettings1>();
            ConfigureAndValidate<MySettings2>();
            ConfigureAndValidate<MySettings3>();

            void ConfigureAndValidate<TOptions>() where TOptions : class
            {
                var config = configuration.GetSection(typeof(TOptions).Name);
                services.AddOptions<TOptions>().Bind(config).ValidateDataAnnotations();
            }
        }
```

### Expected behavior

No S1172 warning is shown for this case.

### Actual behavior

The S1172 warning is shown even though the parameter is required.

### Known workarounds

Suppress the warning, assign the parameter to a discard, or disable the SonarLint extension.

### Related information

* C#/VB.NET Plugins version 4.36.034265
* SonarAnalyzer.CSharp version 8.26.034506
* Visual Studio version 16.10.1
* dotnet version 5.0.302
* Operating System Windows 10 Pro 21H1

### PR DESCRIPTION
**Pull request by pavel-mikula-sonarsource**
Fixes #4704

The key idea is to handle `GenericNameSyntax` (that has `.Identifier`) the same way as `ProcessIdentifier` was updated in #3818

### CODE REVIEW

File: analyzers/src/SonarAnalyzer.CSharp/LiveVariableAnalysis/CSharp/CSharpLiveVariableAnalysis.cs [hunk 1]
```diff
@@ -178,6 +182,16 @@ private void ProcessIdentifier(IdentifierNameSyntax identifier, State state)
             }
         }
 
+        private void ProcessGenericName(SyntaxNode genericName, State state)
+        {
+            if (genericName.Parent.IsKind(SyntaxKind.InvocationExpression)
```
Comment (pavel-mikula-sonarsource) [inline analyzers/src/SonarAnalyzer.CSharp/LiveVariableAnalysis/CSharp/CSharpLiveVariableAnalysis.cs:187]: This is the same logic as in the `ProcessIdentifier(..)` above. We just don't need to care about the `if (IsLocalScoped(..` here.

File: analyzers/src/SonarAnalyzer.CSharp/LiveVariableAnalysis/CSharp/CSharpLiveVariableAnalysis.cs [hunk 2]
```diff
@@ -178,6 +182,16 @@ private void ProcessIdentifier(IdentifierNameSyntax identifier, State state)
...
+                && semanticModel.GetSymbolInfo(genericName).Symbol is IMethodSymbol method
```
Comment (costin-zaharia-sonarsource) [inline analyzers/src/SonarAnalyzer.CSharp/LiveVariableAnalysis/CSharp/CSharpLiveVariableAnalysis.cs:188]: Can be rewritten like: `&& semanticModel.GetSymbolInfo(genericName).Symbol is IMethodSymbol { MethodKind: MethodKindEx.LocalFunction } method`

**Review by costin-zaharia-sonarsource — CHANGES_REQUESTED** [review]
I think the solution looks good so far but the implementation is not yet complete.

Can you please add the following test case?
```
private static void M(this int someNumber, string someString)
{
    Action x = PrintSomeSum<int>;

    x();

    void PrintSomeSum<TOptions>() where TOptions : struct
    {
        Console.WriteLine(someNumber + someString.Length);
    }
}
```

**Review by costin-zaharia-sonarsource — APPROVED** [review]
LGTM!

**Pull request merged**